src/main/java/com/puppycrawl/tools/checkstyle/site/XdocsTemplateParser.java

```diff
@@ -0,0 +1,204 @@
+     * Afterward, the macro fields are reinitialized.
+     *
+     * @param sink the sink object.
+     * @throws MacroExecutionException if a macro is not found.
+     */
+    private void processMacroEnd(Sink sink) throws MacroExecutionException {
+        if (macroName != null && !macroName.isEmpty() && !isSecondParsing()) {
+            final MacroRequest request = new MacroRequest(sourceContent,
+                    new XdocsTemplateParser(), macroParameters,
+                    new File(System.getProperty("user.dir")));
+
+            try {
+                executeMacro(macroName, request, sink);
+            }
+            catch (MacroNotFoundException exception) {
+                throw new MacroExecutionException("Macro not found: " + macroName, exception);
+            }
+        }
+
+        reinitializeMacroFields();
+    }
+
+    /**
+     * Reinitialize the macro fields.
+     */
+    private void reinitializeMacroFields() {
+        macroName = null;
+        macroParameters = null;
+    }
+
+    /**
+     * Handle the opening tag of a macro. Gather the macro name and parameters.
+     *
+     * @param parser the xml parser.
+     * @throws MacroExecutionException if the macro name is not specified.
+     */
+    private void processMacroStart(XmlPullParser parser) throws MacroExecutionException {
+        if (!isSecondParsing()) {
+            macroName = parser.getAttributeValue(null, Attribute.NAME.toString());
+
+            if (macroParameters == null) {
+                macroParameters = new HashMap<>();
+            }
+
+            if (macroName == null || macroName.isEmpty()) {
+                throw new MacroExecutionException("The '"
+                        + Attribute.NAME
+                        + "' attribute for the '"
+                        + MACRO_TAG + "' tag is required.");
+            }
+        }
+    }
+
+    /**
+     * Handle the opening tag of a parameter. Gather the parameter name and value.
+     *
+     * @param parser the xml parser.
+     * @param sink the sink object.
+     * @throws MacroExecutionException if the parameter name or value is not specified.
+     */
+    private void processParamStart(XmlPullParser parser, Sink sink) throws MacroExecutionException {
+        if (!isSecondParsing()) {
+            if (macroName != null && !macroName.isEmpty()) {
+                final String paramName = parser
+                        .getAttributeValue(null, Attribute.NAME.toString());
+                final String paramValue = parser
+                        .getAttributeValue(null, Attribute.VALUE.toString());
+
+                if (paramName == null
+                        || paramValue == null
+                        || paramName.isEmpty()
+                        || paramValue.isEmpty()) {
+                    throw new MacroExecutionException(
+                            "'" + Attribute.NAME + "' and '" + Attribute.VALUE
```

Comment:
Please use `String.format` for message generation.

Reply:
Done.